tutorials/holoscan_debugging/README.md

```diff
@@ -0,0 +1,286 @@
+```sh
+./tutorials/holoscan_debugging/debug_pdb.sh
+```
+
+The script runs through the following steps:
+1. Builds the HoloHub container environment based on the Holoscan SDK NGC container.
+2. Builds the Endoscopy Tool Tracking __Python__ application in the container environment. By default we use the
+  `debug` build mode to generate detailed C++ debugging symbols for the Endoscopy Tool Tracking application.
+3. Launches the Endoscopy Tool Tracking application with `pdb`, which is invoked via the Python interpreter:
+```bash
+python3 -m pdb <command --args ...>
+```
+
+This command launches the Python version of the Endoscopy Tool Tracking application with a breakpoint set on the
+very first line. From this point you can set additional breakpoints, inspect application state, and control program
+execution.
+
+Although launched from Python, calls to `import holoscan` will load Python-wrapped C++ Holoscan SDK libraries
+along with their debugging symbols. You can set breakpoints in the corresponding Holoscan SDK C++ source code after each
+breakpoint launches. For instance:
+```py
+# Load the Holoscan core module in Python
+-> from holoscan.core import Application
+(Pdb) n
+> /workspace/holohub/applications/endoscopy_tool_tracking/python/endoscopy_tool_tracking.py(20)<module>()
+# Core debugging symbols are now available we can set breakpoints in C++ code.
+(Pdb) break /workspace/holoscan-sdk/src/core/application.cpp:172
+Breakpoint 1 at /workspace/holoscan-sdk/src/core/application.cpp:172
+...
+```
+
+You can also add a breakpoint directly in the `.py` source code before running a program by adding a `breakpoint()` statement.
+
+From here we recommend referring to [`pdb` documentation](https://docs.python.org/3/library/pdb.html) for common
+commands and debugging strategies.
+
+## Debugging Symbols for Legacy Holoscan SDK Versions
+
+In Holoscan SDK v2.3 we began packaging debugging symbols as part of the libraries distributed in
+[NGC Holoscan Containers](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan), with the goal
+of improving ease of development and debugging. This change comes at a small cost to size and performance of the Holoscan SDK
+binary distribution. But what about older versions of Holoscan SDK containers?
+
+If you are using a legacy Holoscan SDK container (earlier than v2.3) for your development, your container does not come with
+debugging symbols pre-packaged. However, you can rebuild Holoscan SDK from source to enable interactive debugging.
+We provide utilities as part of Holoscan SDK open source code to help you generate debugging
+symbols that will allow you to use tools such as `gdb` and `pdb` with legacy Holoscan SDK.
+
+The following script provides the necessary steps to rebuild Holoscan SDK version with debugging symbols and then
+set up for debugging with `gdb`:
+
+```bash
+./tutorials/holoscan_debugging/debug_legacy.sh <build_type> <sdk_tag> <ngc_tag>
```

Comment:
Can we provide an example with actual values? :pray:  
Maybe for convenience, links to the tags and NGC page can also be added so the user can easily locate available tags.

https://github.com/nvidia-holoscan/holoscan-sdk/tags
https://catalog.ngc.nvidia.com/orgs/nvidia/teams/clara-holoscan/containers/holoscan/tags

Reply:
The script uses default values for v2.2.0. Would it be more helpful to remove the generic arguments here or to replace them with the default values?

Reply:
I think what you have looks good, but providing an example would make it easier to follow because not everyone knows what `ngc_tag` means. A few links to let the users know where to find the values would help, too.

Reply:
Due to https://github.com/nvidia-holoscan/holoscan-sdk/issues/30, the only legacy version that can be patched and built here is v2.2.0. Refactored the script and removed these arguments accordingly. (TBD whether we remove the legacy script as well.)